xds/internal/balancer/wrrlocality/balancer.go

```diff
@@ -45,24 +49,153 @@ func (bb) Name() string {
-		return nil, errors.New("xds: invalidw LBConfig for wrrlocality: child policy field must be set")
+		return nil, errors.New("xds: invalid LBConfig for wrrlocality: child policy field must be set")
 	}
 	return lbCfg, nil
 }
+
+type attributeKey struct{}
+
+// Equal allows the values to be compared by Attributes.Equal.
+func (a AddrInfo) Equal(o interface{}) bool {
+	oa, ok := o.(AddrInfo)
+	return ok && oa.LocalityWeight == a.LocalityWeight
+}
+
+// AddrInfo is the locality weight of the locality an address is a part of.
+type AddrInfo struct {
+	LocalityWeight uint32
+}
+
+// SetAddrInfo returns a copy of addr in which the BalancerAttributes field is
+// updated with AddrInfo.
+func SetAddrInfo(addr resolver.Address, addrInfo AddrInfo) resolver.Address {
+	addr.BalancerAttributes = addr.BalancerAttributes.WithValue(attributeKey{}, addrInfo)
+	return addr
+}
+
+func (a AddrInfo) String() string {
+	return fmt.Sprintf("Locality Weight: %d", a.LocalityWeight)
+}
+
+// getAddrInfo returns the AddrInfo stored in the BalancerAttributes field of
+// addr. Returns false if no AddrInfo found.
+func getAddrInfo(addr resolver.Address) (AddrInfo, bool) {
+	v := addr.BalancerAttributes.Value(attributeKey{})
+	ai, ok := v.(AddrInfo)
+	return ai, ok
+}
+
+// wrrLocalityBalancer wraps a weighted target balancer, and builds
+// configuration for the weighted target once it receives configuration
+// specifying the weighted target child balancer and locality weight
+// information.
+type wrrLocalityBalancer struct {
+	// child will be a weighted target balancer, and will be built it at
+	// wrrLocalityBalancer build time. Other than preparing configuration, other
+	// balancer operations are simply pass through.
+	child balancer.Balancer
+
+	childParser balancer.ConfigParser
+
+	logger *grpclog.PrefixLogger
+}
+
+func (b *wrrLocalityBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
+	lbCfg, ok := s.BalancerConfig.(*LBConfig)
+	if !ok {
+		b.logger.Errorf("Received config with unexpected type %T: %v", s.BalancerConfig, s.BalancerConfig)
+		return balancer.ErrBadResolverState
+	}
+
+	weightedTargets := make(map[string]weightedtarget.Target)
+	for _, addr := range s.ResolverState.Addresses {
+		// This get of LocalityID could potentially return a zero value. This
+		// shouldn't happen though (this attribute that is set actually gets
+		// used to build localities in the first place), and thus don't error
+		// out, and just build a weighted target with undefined behavior.
+		locality, err := internal.GetLocalityID(addr).ToString()
+		if err != nil {
+			// Should never happen.
+			logger.Errorf("Failed to marshal LocalityID: %v, skipping this locality in weighted target")
+		}
+		ai, ok := getAddrInfo(addr)
+		if !ok {
+			return fmt.Errorf("addr: %v is misisng locality weight information", addr)
+		}
+		weightedTargets[locality] = weightedtarget.Target{Weight: ai.LocalityWeight, ChildPolicy: lbCfg.ChildPolicy}
```

Comment:
I see the following in the gRFC:
```
Since an xDS configuration can place a given locality under multiple priorities, it is possible
to see locality weight attributes with different values for the same locality. We do not support
this kind of an edge case and just use the weight in the first attribute we encounter. A warning
should be logged about any locality weights being discarded.
```

Should this be handled here or is this handled elsewhere?

Reply:
Yes. I thought about this for a long time and tried to figure out where that line maps (it was part of cluster resolver). However, I realized that it's handled in our codebase at the Parsing of EDS level: https://github.com/grpc/grpc-go/pull/6224/files#diff-83cdefbe0b89dcacebd9706dfe9e03b537d079248eb4cdbe4f6457ca9d36f939R144. This means that each priority sends down it's specific locality weight and this line doesn't apply to our codebase.

Reply:
Ack. Thanks.